### PR TITLE
Add List command to display a list of available exercises

### LINKED ISSUE #89: Display list of all the exercises
```bash
$ exercism list go
```

Should show a list of all the exercises, and explain how to get specific ones.

### PR DESCRIPTION
**Pull request by lcowell**
WIP - I just realize that I need to handle a 500 response in the case that someone tries to list for a language like 'rub' or 'haskll'.

Closes #89

### CODE REVIEW

File: cmd/list.go [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+	"io"
+	"log"
+	"os"
+
+	"github.com/codegangsta/cli"
+	"github.com/exercism/cli/api"
+	"github.com/exercism/cli/config"
+)
+
+const msgExplainFetch = "In order to fetch a specific assignment, call the fetch command with a specific assignment.\n\nexercism fetch ruby/matrix"
```
Comment (kytrinyx) [inline cmd/list.go]: The fetch command actually takes two arguments: `exercism fetch ruby matrix`

File: cmd/list.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+	problems, err := api.List(language, c.XAPI)
+	if err != nil {
+		log.Fatal(err)
+	}
+	GenerateOutput(problems, os.Stdout)
+}
+
+// Output writes out the list of assignments
+func GenerateOutput(problems []string, w io.Writer) {
```
Comment (kytrinyx) [inline cmd/list.go]: I don't really like exposing a function like `GenerateOutput` to the entire `cmd` package. What do you think of adding a new package that can hold on to track-related things? Earlier I've isolated some things into the `Homework` type in the `user` package. Would that approach work?

File: cmd/list.go [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+	for _, p := range problems {
+		fmt.Fprintf(w, "%s\n", p)
```
Comment (kytrinyx) [inline cmd/list.go]: Should we strip off the `ruby/` prefix?

**Comment by kytrinyx** [comment]
It's cool to see progress again on this project! :sparkles: :heart: :sparkles:

**Comment by kytrinyx** [comment]
I think the failing test was due to a stupid mistake I made on master.

File: cmd/list.go (hunk 1 above)
Comment (lcowell) [inline cmd/list.go]: Good catch. It actually worked for me to say `exercism fetch ruby/matrix`, but we should be telling people the right thing to do, not the thing that happens to work in this release.

File: cmd/list.go (hunk 3 above)
Comment (lcowell) [inline cmd/list.go]: Definitely, that would be easier to copy and paste.

File: cmd/list.go (hunk 2 above)
Comment (lcowell) [inline cmd/list.go]: I think I was having a ruby moment here. Of course `GenerateOutput` is going to be available to everything in the `cmd` package. The motivation behind adding the function was to allow me to easily test the generated output. However, the test is quite brittle and adds little value.

My leaning would be towards:
* changing the function name to be something unexported and more appropriate like: `outputProblemList` -or-
* folding the function in to `cmd.List` and only abstracting as required

I'd favour eliminating the function for now. What do you think about doing this for now and exploring the value of a `track` package once we merge this in ?

**Comment by lcowell** [comment]
One last thing I was wondering about is: should we be generating a better response when someone requests a language that doesn't exist ? eg `http://x.exercism.io/tracks/rby` will just give me a 500. That's ok, but I can't really know if I'm getting a 500 because of the language or because there's something going on with the server.

File: cmd/list.go (hunk 2 above)
Comment (kytrinyx) [inline cmd/list.go]: Yeah, I think inlining it is good. We can refactor later if necessary.

**Comment by kytrinyx** [comment]
> http://x.exercism.io/tracks/rby

That should return a 404, not a 500.

https://github.com/exercism/x-api/issues/37

File: api/api.go [hunk 4]
```diff
@@ -128,6 +128,31 @@ func Submit(url string, iter *Iteration) (*Submission, error) {
 	return ps.Submission, nil
 }
 
+// List available problems for a language
+func List(language, host string) ([]string, error) {
```
Comment (Tonkpils) [inline api/api.go]: I like the way this function takes separate arguments as opposed to the composed url. Even providing host seems a bit off but until we have a Client struct for the api package this is fine. I think we should follow this same pattern for the other functions in this file. Any thoughts on this?

Comment (lcowell) [inline api/api.go]: Thanks! I think what I like about this is that the caller of `api.List` doesn't need to know about how to compose a URL. AFAIK each of the api commands takes a url as an argument which means those URLs are probably being generated in each of the `cmd` funcs. I don't think that `cmd` package should know about URLs.

What do you think about making functions for generating URLs and putting them in the api package ? It kind of seems like the right place for them to me.

```go
  func ListURL(language, host string) string {}
  func UnsubmitURL(host string) string {}
  func FetchURL(parts ...string) string {}
```

Comment (kytrinyx) [inline api/api.go]: Yeah, agreed. Let's let the api worry about urls (functions or hard-coded), and the function arguments can be whatever information is needed.

Comment (kytrinyx) [inline api/api.go]: That said, that is probably an unrelated refactoring.

**Comment by kytrinyx** [comment]
The API now returns a 404 for non-existent languages.

Do you want to update the PR with this before we merge? Once you're ready would you mind squashing the commits?

**Comment by lcowell** [comment]
Handling 404s is in place. Commits squashed. Wahoo!

**Comment by kytrinyx** [comment]
Sweeeet!
